Readme examples (#1425)
linera-io/linera-protocol: examples/amm/README.md

<!-- cargo-rdme start -->

# Automated Market Maker (AMM) Example Application

This example implements an Automated Market Maker (AMM) which demonstrates DeFi capabilities of the 

> Linera protocol. Pre-requisites for the AMM application is the `funglible` application, as we will

Reviewer: ```suggestion
Linera protocol. Prerequisite for the AMM application is the `fungible` application, as we will
```

be adding, removing liquidity and also perform a swap.

# How it works 

> It supports the following operations

Reviewer: ```suggestion
It supports the following operations.
```

- Swap: For a given input token and an input amount, it swaps that token amount for an 
amount of the other token calculated based on the current AMM ratio. Note: The `Swap` operations 

> are needed to performed from a remote chain 

Reviewer 2: ```suggestion
need to be performed from a remote chain 
```

- Add Liquidity: This operation allows adding liquidity to the AMM. Given a maximum 
`token0` and `token1` amount that you're willing to add, it adds liquidity such that you'll be 
adding at most `max_token0_amount` of `token0` and `max_token1_amount` of `token1`. The amounts 
will be calculated based on the current AMM ratio. The owner, in this case, refers to the user 
adding liquidity, which currently can only be a chain owner.

> - Remove Liquidity: This is a liquidity removal operation. Given a token idx of the token you'd 

Reviewer: (The first sentence doesn't explain much; maybe:)

```suggestion
- Remove Liquidity: This withdraws tokens from the AMM. Given the index of the token you'd 
```

like to remove (can be 0 or 1), and an amount of that token that you'd like to remove, it calculates 
how much of the other token will also be removed based on the current AMM ratio. Then it removes 
the amounts from both tokens as a removal of liquidity. The owner, in this context, is the user 
removing liquidity, which currently can only be a chain owner.

# Usage

## Setting Up

First, build Linera and add it to the path:

```bash
cargo build
export PATH=$PWD/target/debug:$PATH
```

> To start the local linera network

Reviewer: ```suggestion
To start the local Linera network
```

```bash
linera net up --testing-prng-seed 37
```

> This will start the local linera network. We used the

Reviewer: ```suggestion
This will start the local Linera network. We used the
```

test-only CLI option `--testing-prng-seed` to make keys deterministic and simplify our
presentation.

Copy and paste `export LINERA_WALLET="/var . . . ."` & `export LINERA_STORAGE="rocksdb:/. . . ."` from output of `linera net up --testing-prng-seed 37` into another terminal

```bash
OWNER_1=e814a7bdae091daf4a110ef5340396998e538c47c6e7d101027a225523985316
OWNER_2=453690095cdfe6dbde7fc577e56bb838a7ee7920a72512d4a87748b4e151ed61
CHAIN_1=e476187f6ddfeb9d588c7b45d3df334d5501d6499b3f9ad5595cae86cce16a65
CHAIN_2=e54bdb17d41d5dbe16418f96b70e44546ccd63e6f3733ae3c192043548998ff3
```

Now we have to publish and deploy the AMM application,

```bash
appid1=YOUR_FUNGIBLE_APP1_ID

> appid2=YOUR_FUNGIBLE_APP2_ID

Reviewer: Aren't the app IDs deterministic, too? I'd include the commands to deploy the fungible apps in this README.

(Is there a reason you use lowercase for these?)

Author: Something like this?

```
FUN1_APP_ID=$(linera publish-and-create ../target/wasm32-unknown-unknown/release/fungible_{contract,service}.wasm \
    --json-argument "{ \"accounts\": {
        \"User:$OWNER_1\": \"100.\",
        \"User:$OWNER_2\": \"150.\"
    } }" \
    --json-parameters "{ \"ticker_symbol\": \"FUN1\" }" \
)
```

Reviewer: Yes, that looks right. :+1:

> (cd examples/counter && cargo build --release)

Reviewer: ```suggestion
(cd examples/amm && cargo build --release)
```

AMM_APPLICATION_ID=(linera publish-and-create ../target/wasm32-unknown-unknown/release/amm_{contract,service}.wasm --json-parameters "{\"tokens\":["\"$appid1\"","\"$appid2\""]}")
```

## Using the Token Application

First, a node service for the current wallet has to be started:

```bash
PORT=8080
linera service --port $PORT &
```

### Using GraphiQL

> Before performing any operation we need to provide liquidity to it, we will `AddLiquidity` operation,

Reviewer: ```suggestion
Before performing any operation we need to provide liquidity to it, so we will use the `AddLiquidity` operation,
```

navigate to `http://localhost:8080/chains/$CHAIN_1/applications/$AMM_APPLICATION_ID`

To perform `AddLiquidity` operation:

```json
mutation{
  operation(
    operation: {
      AddLiquidity: {
        owner:"User:e814a7bdae091daf4a110ef5340396998e538c47c6e7d101027a225523985316",
        max_token0_amount: "50",
        max_token1_amount: "40",
      }
    } 		
  )
}
```

We can only perform `Swap` from a remote chain i.e. other than the chain on which  `AMM` is deployed to,
we can do it from GraphiQL by perfoming a mutation `requestApplication` so that we can perform the `Swap` operation from the
chain

```json
mutation {
  requestApplication (
    chainId:"e54bdb17d41d5dbe16418f96b70e44546ccd63e6f3733ae3c192043548998ff3",
    applicationId: "AMM_APPLICATION_ID",
    targetChainId: "e476187f6ddfeb9d588c7b45d3df334d5501d6499b3f9ad5595cae86cce16a65"
  )
}
```

> Note: Above mutation has to perform from `http://localhost:8080`

Reviewer: ```suggestion
Note: The above mutation has to be performed from `http://localhost:8080`
```

Now to perform `Swap` operation, naviage to `http://localhost:8080/chains/$CHAIN_2/applications/$AMM_APPLICATION_ID` and
perform the following mutation

```json
mutation{
  operation(
    operation: {
      Swap: {
        owner:"User:453690095cdfe6dbde7fc577e56bb838a7ee7920a72512d4a87748b4e151ed61",
        input_token_idx: 1,
        input_amount: "1",
      }

>     } 		

Reviewer: ```suggestion
    }
```

  )
}
```

We can also `RemoveLiquidiy` operation, navigate to `http://localhost:8080/chains/$CHAIN_1/applications/$AMM_APPLICATION_ID` and
perform the following mutation

```json
mutation{
  operation(
    operation: {
      RemoveLiquidity: {
        owner:"User:453690095cdfe6dbde7fc577e56bb838a7ee7920a72512d4a87748b4e151ed61",
        token_to_remove_idx: 1,
        token_to_remove_amount: "1",
      }

>     } 		

Reviewer: ```suggestion
    }
```

<!-- cargo-rdme start -->

# Automated Market Maker (AMM) Example Application

This example implements an Automated Market Maker (AMM) which demonstrates DeFi capabilities of the 

> Linera protocol. Prerequisite for the AMM application is the `fungilble` application, as we will

Reviewer: ```suggestion
Linera protocol. Prerequisite for the AMM application is the `fungible` application, as we will
```

`token0` and `token1` amount that you're willing to add, it adds liquidity such that you'll be 
adding at most `max_token0_amount` of `token0` and `max_token1_amount` of `token1`. The amounts 
will be calculated based on the current AMM ratio. The owner, in this case, refers to the user 
adding liquidity, which currently can only be a chain owner.

- Remove Liquidity: This withdraws tokens from the AMM. Given the index of the token you'd 
like to remove (can be 0 or 1), and an amount of that token that you'd like to remove, it calculates 
how much of the other token will also be removed based on the current AMM ratio. Then it removes 
the amounts from both tokens as a removal of liquidity. The owner, in this context, is the user 
removing liquidity, which currently can only be a chain owner.

# Usage

## Setting Up

First, build Linera and add it to the path:

```bash
cargo build
export PATH=$PWD/target/debug:$PATH
```

To start the local Linera network

```bash
linera net up --testing-prng-seed 37
```

This will start the local Linera network. We used the
test-only CLI option `--testing-prng-seed` to make keys deterministic and simplify our

> presentation.

Reviewer 2: ```suggestion
explanation.
```

Copy and paste `export LINERA_WALLET="/var . . . ."` & `export LINERA_STORAGE="rocksdb:/. . . ."` from output of `linera net up --testing-prng-seed 37` into another terminal

```bash
OWNER_1=e814a7bdae091daf4a110ef5340396998e538c47c6e7d101027a225523985316
OWNER_2=453690095cdfe6dbde7fc577e56bb838a7ee7920a72512d4a87748b4e151ed61
CHAIN_1=e476187f6ddfeb9d588c7b45d3df334d5501d6499b3f9ad5595cae86cce16a65
CHAIN_2=e54bdb17d41d5dbe16418f96b70e44546ccd63e6f3733ae3c192043548998ff3
```

> Now we have to publish and create the fungible applications,

Reviewer 2: ```suggestion
Now we have to publish and create the fungible applications:
```

    --json-argument "{ \"accounts\": {
        \"User:$OWNER_1\": \"100.\",
        \"User:$OWNER_2\": \"150.\"
    } }" \
    --json-parameters "{ \"ticker_symbol\": \"FUN2\" }" \
)

(cd examples/amm && cargo build --release)
AMM_APPLICATION_ID=$(linera publish-and-create ./examples/target/wasm32-unknown-unknown/release/amm_{contract,service}.wasm --json-parameters "{\"tokens\":["\"$FUN1_APP_ID\"","\"$FUN2_APP_ID\""]}")
```

## Using the AMM Application

First, a node service for the current wallet has to be started:

```bash
PORT=8080
linera service --port $PORT &
```

### Using GraphiQL

Before performing any operation we need to provide liquidity to it, so we will use the `AddLiquidity` operation,
navigate to `http://localhost:8080/chains/$CHAIN_1/applications/$AMM_APPLICATION_ID`

To perform `AddLiquidity` operation:

```json
mutation{
  operation(
    operation: {
      AddLiquidity: {
        owner:"User:e814a7bdae091daf4a110ef5340396998e538c47c6e7d101027a225523985316",
        max_token0_amount: "50",
        max_token1_amount: "40",
      }
    }
  )
}
```

> We can only perform `Swap` from a remote chain i.e. other than the chain on which  `AMM` is deployed to,

Reviewer 2: ```suggestion
We can only perform `Swap` from a remote chain i.e. other than the chain on which `AMM` is deployed to,
```

> we can do it from GraphiQL by perfoming a mutation `requestApplication` so that we can perform the `Swap` operation from the

Reviewer 2: ```suggestion
we can do it from GraphiQL by perfoming the `requestApplication` mutation so that we can perform the `Swap` operation from the
```

chain

```json
mutation {
  requestApplication (
    chainId:"e54bdb17d41d5dbe16418f96b70e44546ccd63e6f3733ae3c192043548998ff3",
    applicationId: "AMM_APPLICATION_ID",
    targetChainId: "e476187f6ddfeb9d588c7b45d3df334d5501d6499b3f9ad5595cae86cce16a65"
  )
}
```
Note: The above mutation has to be performed from `http://localhost:8080`

Now to perform `Swap` operation, naviage to `http://localhost:8080/chains/$CHAIN_2/applications/$AMM_APPLICATION_ID` and

> perform the following mutation

Reviewer 2: ```suggestion
perform the following mutation:
```

```json
mutation{
  operation(
    operation: {
      Swap: {
        owner:"User:453690095cdfe6dbde7fc577e56bb838a7ee7920a72512d4a87748b4e151ed61",
        input_token_idx: 1,
        input_amount: "1",
      }
    }
  )
}
```

> We can also `RemoveLiquidiy` operation, navigate to `http://localhost:8080/chains/$CHAIN_1/applications/$AMM_APPLICATION_ID` and

Reviewer 2: ```suggestion
We can also perform the `RemoveLiquidity` operation, navigate to `http://localhost:8080/chains/$CHAIN_1/applications/$AMM_APPLICATION_ID` and
```

> perform the following mutation

Reviewer 2: ```suggestion
perform the following mutation:
```
